metric_learn/constraints.py

```diff
@@ -33,6 +35,104 @@ def positive_negative_pairs(self, num_constraints, same_length=False,
       return a[:n], b[:n], c[:n], d[:n]
     return a, b, c, d
 
+  def generate_knntriplets(self, X, k_genuine, k_impostor):
+    """
+    Generates triplets for every point to `k_genuine` neighbors of the same
+    class and `k_impostor` neighbors of other classes.
+
+    For every point (X_a) the triplets (X_a, X_b, X_c) are constructed from all
+    the combinations of taking `k_genuine` neighbors (X_b) of the same class
+    and `k_impostor` neighbors (X_c) of other classes.
+
+    Parameters
+    ----------
+      X : (n x d) matrix
+        Input data, where each row corresponds to a single instance.
+      k_genuine : int
+        Number of neighbors of the same class to be taken into account.
+      k_impostor : int
+        Number of neighbors of different classes to be taken into account.
+
+    Returns
+    -------
+    triplets : array-like, shape=(n_constraints, 3)
+      2D array of triplets of indicators.
+    """
+
+    labels, labels_count = np.unique(self.partial_labels, return_counts=True)
+    n_labels = len(labels)
+    len_input = np.size(self.partial_labels, 0)
+
+    # Handle the case where there are too few elements to yield k_genuine or
+    # k_impostor neighbors for every class.
+
+    k_genuine_vec = np.ones(n_labels, dtype=np.intp)*k_genuine
+    k_impostor_vec = np.ones(n_labels, dtype=np.intp)*k_impostor
+
+    for i in range(n_labels):
+      if (k_genuine + 1 > labels_count[i]):
+        k_genuine_vec[i] = labels_count[i]-1
+        warnings.warn("The class {} has {} elements but a minimum of {},"
+                      " which corresponds to k_genuine+1, is expected. "
+                      "A lower number of k_genuine will be used for this"
+                      "class.\n"
```

Comment:
This could be more direct and informative:
The class {} has {} elements, which is not sufficient to generate {} genuine neighbors as specified by k_genuine. Will generate {} genuine neighbors instead.